torch/csrc/jit/script/final_returns.cpp

```diff
@@ -1,106 +1,288 @@
-#include <torch/csrc/jit/ir.h>
 #include <torch/csrc/jit/script/final_returns.h>
+#include <ATen/core/jit_type.h>
+#include <torch/csrc/jit/ir.h>
+#include <torch/csrc/jit/ir_views.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+#include <torch/csrc/jit/script/error_report.h>
 
 namespace torch {
 namespace jit {
 namespace script {
 
-struct ReturnInfo {
-  bool returns_; // true - all paths through stmts_ always return
-                 // false - all paths through stmts_ do not return
-  List<Stmt> stmts_;
-};
+/**
+ * This pass transforms the Graph so that all ReturnStmts are merged into a
+ * single value at the end of the graph.
+ *
+ * For blocks and control flow nodes that have a return statement that may have
+ * been hit, we add an extra output for the return value, and an extra output
+ * indicating whether or not the return has been hit (a sentinel value).
+ *
+ * When we encounter a node that might return, we guard all subsequent nodes
+ * in the block with the sentinel value of that node.
+ */
+
+// Will a block or node return
+enum Return_Status { WONT_RETURN, MIGHT_RETURN, WILL_RETURN };
+
+// The return output of Control Flow Nodes is the second to last output,
+// sentinel output is the output
+constexpr size_t RETURN_OFFSET = 2;
```

Comment:
This makes it seem like _every_ control node will have both these outputs. But only some will.